rfcs/0009-abstraction-layer-for-mobile-error-tracking.md

```diff
@@ -0,0 +1,309 @@
+
+const errorTracking = useErrorTracking();
+
+const aFunctionThatMightFail = () => {
+  throw new Error("Failed");
+};
+
+const Button = () => {
+  const onPress = useCallback(() => {
+    try {
+      aFunctionThatMightFail();
+    } catch (err) {
+      errorTracking.captureError(err);
+    }
+  }, []);
+
+  return (
+    <TouchableOpacity onPress={onPress}>
+      <Text>Press Here</Text>
+    </TouchableOpacity>
+  );
+};
+```
+
+Note that it's possible to enable or disable the error tracking as you wish, so it would be possible to:
+
+```js
+import React, { useState, useCallback } from "react";
+import { View, ouchableOpacity, Text } from "react-native";
+import createErrorTracking from "@moxy/react-native-sentry";
+import { useErrorTracking } from "../hooks/error-tracking";
+
+const ToggleErrorTracking = () => {
+  const [enabled, setEnabled] = useState(true);
+
+  const errorTracking = useErrorTracking();
+
+  const onPress = useCallback(() => {
+    setEnabled((state) => !state);
+
+    errorTracking.enable(enabled);
+  }, [setEnabled, errorTracking]);
+
+  return (
+    <View>
+      <TouchableOpacity onPress={onPress}>
+        <Text>Toggle Error Tracking</Text>
+      </TouchableOpacity>
+
+      <Text>Error tracking is {enabled ? "enabled" : "disabled"}</Text>
+    </View>
+  );
+};
+```
+
+## Regarding the initialization
+
+In my opinion the initialization should not be abstracted, since Sentry and Firebase Crashlytics have quite differente ways of being initialized.
+
+- Sentry:
+
+```js
+import * as Sentry from '@sentry/react-native';
+
+Sentry.init({
+    dsn: // Sentry's DSN,
+});
+```
+
+- Firebase Crashlytics:
+
+```js
+import crashlytics from "@react-native-firebase/crashlytics";
+
+// you can already use it inside the file
+```
+
+# Drawbacks
+
+The only drawback I can see is the initial effort to design the solution, since the maintenance should be really simple.
+
+# Alternatives
+
+The alternative would be using the API provided directly from the error tracking solutions. However, this is not a good idea as I've already commented about in the "Motivation" section.
+
+# Adoption strategy
+
+To integrate those modules `@moxy/react-native-{solution}`, where `solution` would be the name of the chosen error tracking solution (it could be `sentry` for instance), one should start by installing it:
+
+```sh
+npm i @moxy/react-native-sentry
+
+# yarn add @moxy/react-native-sentry
+```
+
+On the `src/app/App.js` file, you should initialize Sentry:
+
+```js
+// src/app/App.js
+import * as Sentry from "@sentry/react-native";
+import createErrorTracking from "@moxy/react-native-sentry";
+import { ErrorTrackingProvider } from "../../shared/hooks/error-tracking";
+
+Sentry.init({
+  dsn: // Sentry's DSN,
+});
+
+const errorTracking = createErrorTracking(Sentry, { enabled: true });
+
+const App = () => (
+  <ErrorTrackingProvider value={errorTracking}>
+    {/* ... */}
+  </ErrorTrackingProvider>
+);
+```
```

Comment:
This is already included in the setup section so there is no need to repeat it here.

```suggestion
```

Reply:
I don't see a problem repeating this here, since this is just there to make it clear how to the initialization should be done, but I will remove it anyways.